apisix/discovery/nacos.lua

```diff
@@ -0,0 +1,316 @@
+
+local default_weight
+local applications
+local auth_path = "auth/login"
+local instance_list_path = "ns/instance/list?healthyOnly=true&serviceName="
+
+local schema = {
+    type = "object",
+    properties = {
+        host = {
+            type = "array",
+            minItems = 1,
+            items = {
+                type = "string",
+            },
+        },
+        fetch_interval = {type = "integer", minimum = 1, default = 30},
+        prefix = {type = "string", default = "/nacos/v1/"},
+        weight = {type = "integer", minimum = 1, default = 100},
+        timeout = {
+            type = "object",
+            properties = {
+                connect = {type = "integer", minimum = 1, default = 2000},
+                send = {type = "integer", minimum = 1, default = 2000},
+                read = {type = "integer", minimum = 1, default = 5000},
+            },
+            default = {
+                connect = 2000,
+                send = 2000,
+                read = 5000,
+            }
+        },
+    },
+    required = {"host"}
+}
+
+
+local _M = {
+    version = 0.1,
+}
+
+
+local function request(request_uri, path, body, method, basic_auth)
+    local url = request_uri .. path
+    log.info("request url:", url)
+    local headers = core.table.new(0, 0)
+    headers['Accept'] = 'application/json'
+
+    if basic_auth then
+        headers['Authorization'] = basic_auth
+    end
+
+    if body and 'table' == type(body) then
+        local err
+        body, err = core.json.encode(body)
+        if not body then
+            return nil, 'invalid body : ' .. err
+        end
+        headers['Content-Type'] = 'application/json'
+    end
+
+    local httpc = http.new()
+    local timeout = local_conf.discovery.nacos.timeout
+    local connect_timeout = timeout.connect
+    local send_timeout = timeout.send
+    local read_timeout = timeout.read
+    log.info("connect_timeout:", connect_timeout, ", send_timeout:", send_timeout,
+             ", read_timeout:", read_timeout)
+    httpc:set_timeouts(connect_timeout, send_timeout, read_timeout)
+    local res, err = httpc:request_uri(url, {
+        method = method,
+        headers = headers,
+        body = body,
+        ssl_verify = true,
+    })
+    if not res then
+        return nil, err
+    end
+
+    if not res.body or res.status ~= 200 then
+        return nil, "status = " .. res.status
+    end
+
+    local json_str = res.body
+    local data, err = core.json.decode(json_str)
+    if not data then
+        return nil, err
+    end
+    return data
+end
+
+
+local function get_url(request_uri, path)
+    return request(request_uri, path, nil, "GET", nil)
+end
+
+
+local function post_url(request_uri, path, body)
+    return request(request_uri, path, body, "POST", nil)
+end
+
+
+local function get_token_param(base_uri, username, password)
+    if username and password then
+        local data, err = post_url(base_uri, auth_path .. "?username=" .. username
```

Comment:
I think we should call [ngx.encode_args](https://github.com/openresty/lua-nginx-module#ngxencode_args) for `username` and `password`